File: SOCKS/Benchmark/TrafficCounters.swift
import Foundation

final class TrafficCounters {
    enum Direction {
        case upload
        case download
    }

    struct Snapshot: Equatable {
        let uploadBytes: UInt64
        let downloadBytes: UInt64
        let activeTCP: Int
        var activeUDP: Int = 0

        var activeTotal: Int { activeTCP + activeUDP }
    }

    private let queue: DispatchQueue
    private let enabled: Bool
    private var uploadBytes: UInt64 = 0
    private var downloadBytes: UInt64 = 0
    private var activeSessions = Set<UUID>()
    private var activeAssociations = Set<UUID>()

    init(queue: DispatchQueue, enabled: Bool) {
        self.queue = queue
        self.enabled = enabled
    }

    func sessionOpened(_ id: UUID) {
        assertQueue()
        guard enabled else { return }
        activeSessions.insert(id)
    }

    func sessionClosed(_ id: UUID) {
        assertQueue()
        guard enabled else { return }
        activeSessions.remove(id)
    }

    /// UDP associations are counted separately: an association is not a
    /// connection, and `activeTCP` stays a CONNECT-only metric.
    func associationOpened(_ id: UUID) {
        assertQueue()
        guard enabled else { return }
        activeAssociations.insert(id)
    }

    func associationClosed(_ id: UUID) {
        assertQueue()
        guard enabled else { return }
        activeAssociations.remove(id)
    }

    func closeAllSessions() {
        assertQueue()
        activeSessions.removeAll(keepingCapacity: true)
        activeAssociations.removeAll(keepingCapacity: true)
    }

    func recordCommitted(_ byteCount: Int, direction: Direction) {
        assertQueue()
        guard enabled, byteCount > 0 else { return }

        let amount = UInt64(byteCount)
        switch direction {
        case .upload:
            uploadBytes = addingWithoutCrash(uploadBytes, amount)
        case .download:
            downloadBytes = addingWithoutCrash(downloadBytes, amount)
        }
    }

    func snapshot() -> Snapshot {
        assertQueue()
        return Snapshot(
            uploadBytes: uploadBytes,
            downloadBytes: downloadBytes,
            activeTCP: activeSessions.count,
            activeUDP: activeAssociations.count
        )
    }

    func snapshot(_ completion: @escaping (Snapshot) -> Void) {
        queue.async {
            completion(self.snapshot())
        }
    }

    private func assertQueue() {
        dispatchPrecondition(condition: .onQueue(queue))
    }

    private func addingWithoutCrash(_ current: UInt64, _ amount: UInt64) -> UInt64 {
        let (sum, overflowed) = current.addingReportingOverflow(amount)
        return overflowed ? UInt64.max : sum
    }
}

File: android/socks-core/src/main/kotlin/com/nanako/socksbypass/core/TrafficCounters.kt
package com.nanako.socksbypass.core

import java.util.UUID
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.ConcurrentHashMap

class TrafficCounters(private val enabled: Boolean = true) {
    enum class Direction { Upload, Download }

    data class Snapshot(
        val uploadBytes: Long,
        val downloadBytes: Long,
        val activeTcp: Int,
        val activeUdp: Int,
    ) {
        val activeTotal: Int get() = activeTcp + activeUdp
    }

    private val uploadBytes = AtomicLong(0)
    private val downloadBytes = AtomicLong(0)
    private val activeSessions = ConcurrentHashMap.newKeySet<UUID>()
    private val activeAssociations = ConcurrentHashMap.newKeySet<UUID>()

    fun sessionOpened(id: UUID) {
        if (!enabled) return
        activeSessions.add(id)
    }

    fun sessionClosed(id: UUID) {
        if (!enabled) return
        activeSessions.remove(id)
    }

    fun associationOpened(id: UUID) {
        if (!enabled) return
        activeAssociations.add(id)
    }

    fun associationClosed(id: UUID) {
        if (!enabled) return
        activeAssociations.remove(id)
    }

    fun closeAllSessions() {
        activeSessions.clear()
        activeAssociations.clear()
    }

    fun recordCommitted(byteCount: Int, direction: Direction) {
        if (!enabled || byteCount <= 0) return
        val amount = byteCount.toLong()
        when (direction) {
            Direction.Upload -> uploadBytes.updateAndGet { addSaturating(it, amount) }
            Direction.Download -> downloadBytes.updateAndGet { addSaturating(it, amount) }
        }
    }

    fun snapshot(): Snapshot = Snapshot(
        uploadBytes = uploadBytes.get(),
        downloadBytes = downloadBytes.get(),
        activeTcp = activeSessions.size,
        activeUdp = activeAssociations.size,
    )

    private fun addSaturating(current: Long, amount: Long): Long {
        val sum = current + amount
        return if (sum < current) Long.MAX_VALUE else sum
    }
}
